src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -845,3 +1014,172 @@ impl<T: fmt::Display + fmt::Debug + DisplayableInTimeline + TimestampManipulatio
         Ok(())
     }
 }
+
+/// Types and logic in support of a constraint-based approach to timestamp determination.
+mod constraints {
+
+    use core::fmt;
+    use std::fmt::Debug;
+
+    use differential_dataflow::lattice::Lattice;
+    use mz_storage_types::sources::Timeline;
+    use serde::{Deserialize, Serialize};
+    use timely::progress::{Antichain, Timestamp};
+
+    use mz_compute_types::ComputeInstanceId;
+    use mz_repr::GlobalId;
+    use mz_sql::session::vars::IsolationLevel;
+
+    use super::DisplayableInTimeline;
+
+    /// Constraints expressed on the timestamp of a query.
+    ///
+    /// The constraints are expressed on the minimum and maximum values,
+    /// resulting in a (possibly empty) interval of valid timestamps.
+    ///
+    /// The constraints may be redundant, in the interest of providing
+    /// more complete explanations, but they may also be minimized at
+    /// any point without altering their behavior by removing redundant
+    /// constraints.
+    ///
+    /// When combined with a `Preference` one can determine an
+    /// ideal timestamp to use.
+    #[derive(Default, Serialize, Deserialize, Clone)]
+    pub struct Constraints {
+        /// Timestamps and reasons that impose an inclusive lower bound.
+        pub lower: Vec<(Antichain<mz_repr::Timestamp>, Reason)>,
+        /// Timestamps and reasons that impose an inclusive upper bound.
+        pub upper: Vec<(Antichain<mz_repr::Timestamp>, Reason)>,
+    }
+
+    impl DisplayableInTimeline for Constraints {
+        fn fmt(&self, timeline: Option<&Timeline>, f: &mut fmt::Formatter) -> fmt::Result {
+            if !self.lower.is_empty() {
+                writeln!(f, "lower:")?;
+                for (ts, reason) in &self.lower {
+                    let ts = ts.iter().map(|t| t.display(timeline)).collect::<Vec<_>>();
+                    writeln!(f, "  {:?}: ({:?})", ts, reason)?;
+                }
+            }
+            if !self.upper.is_empty() {
+                writeln!(f, "upper:")?;
+                for (ts, reason) in &self.upper {
+                    let ts = ts.iter().map(|t| t.display(timeline)).collect::<Vec<_>>();
+                    writeln!(f, "  {:?}: ({:?})", ts, reason)?;
+                }
+            }
+            Ok(())
+        }
+    }
+
+    impl Debug for Constraints {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            self.display(None).fmt(f)?;
+            Ok(())
+        }
+    }
+
+    impl Constraints {
+        /// Remove constraints that are dominated by other constraints.
+        ///
+        /// This removes redundant constraints, without removing constraints
+        /// that are "tight" in the sense that the interval will not improve
+        /// without their removal. For example, two constraints at the same
```

Comment:
could be misunderstanding, but shouldn't this be "will not improve **with** their removal"?